config/config_graph.go

```diff
@@ -0,0 +1,566 @@
+	if vertex.Type == global && !evaluateGlobals {
+		return false
+	}
+
+	if vertex.Evaluated {
+		return true
+	}
+
+	valuesCty, err := eval.convertValuesToVariables()
+	if err != nil {
+		// TODO: diags.Extend(??)
+		return false
+	}
+
+	ctx := hcl.EvalContext{
+		Variables: valuesCty,
+	}
+
+	value, currentDiags := vertex.Expr.Value(&ctx)
+	if currentDiags != nil && currentDiags.HasErrors() {
+		_ = diags.Extend(currentDiags)
+		return false
+	}
+
+	vertex.Evaluated = true
+
+	switch vertex.Type {
+	case global:
+		eval.globals.values[vertex.Name] = value
+
+	case local:
+		eval.localValues[vertex.Name] = value
+	case include:
+		includePath, includeValue, err := eval.evaluateInclude(value)
+		if err != nil {
+			// TODO: diags.Extend(??)
+			return false
+		}
+
+		eval.includePath = &includePath
+		eval.includeValue = &includeValue
+	default:
+		// TODO: diags.Extend(??)
+		return false
+	}
+
+	return true
+}
+
+func (eval *configEvaluator) evaluateInclude(value cty.Value) (string, map[string]cty.Value, error) {
+	// TODO: validate this is a string?
+	includePath := value.AsString()
+	configPath := eval.configPath
+
+	if includePath == "" {
+		return "", nil, IncludedConfigMissingPath(configPath)
+	}
+
+	childConfigPathAbs, err := filepath.Abs(configPath)
+	if err != nil {
+		return "", nil, err
+	}
+
+	if !filepath.IsAbs(includePath) {
+		includePath = util.JoinPath(filepath.Dir(childConfigPathAbs), includePath)
+	}
+
+	relative, err := util.GetPathRelativeTo(filepath.Dir(configPath), filepath.Dir(includePath))
+	if err != nil {
+		return "", nil, err
+	}
+
+	includeValue := map[string]cty.Value{
+		"parent": cty.StringVal(filepath.ToSlash(filepath.Dir(includePath))),
+		"relative": cty.StringVal(relative),
+	}
+
+	return includePath, includeValue, nil
+}
+
+func (eval *configEvaluator) getBlocks(file *hcl.File) (hcl.Body, hcl.Body, hcl.Body, hcl.Diagnostics) {
+	const locals = "locals"
+	const globals = "globals"
+
+	blocksSchema := &hcl.BodySchema{
+		Blocks: []hcl.BlockHeaderSchema{
+			{Type: locals},
+			{Type: globals},
+			{Type: include},
+		},
+	}
+
+	parsedBlocks, _, diags := file.Body.PartialContent(blocksSchema)
+	if diags != nil && diags.HasErrors() {
+		return nil, nil, nil, diags
+	}
+	blocksByType := map[string][]*hcl.Block{}
+
+	for _, block := range parsedBlocks.Blocks {
+		if block.Type == locals || block.Type == globals || block.Type == include {
+			blocks := blocksByType[block.Type]
+			if blocks == nil {
+				blocks = []*hcl.Block{}
+			}
+
+			blocksByType[block.Type] = append(blocks, block)
+		}
+	}
+
+	// TODO: validate blocks
+
+
+	if _, exists := blocksByType[include]; exists {
+		return blocksByType[locals][0].Body, blocksByType[globals][0].Body, blocksByType[include][0].Body, diags
+	} else {
+		return blocksByType[locals][0].Body, blocksByType[globals][0].Body, nil, diags
+	}
+}
+
+func (eval *configEvaluator) addVertices(vertexType string, block hcl.Body, consumer func(vertex variableVertex) error) error {
```

Comment:
I'm having trouble understanding what this function is doing. I can tell that it is a routine to add vertices to the graph, but what is difficult to understand are:

- Each of the conditions for when a vertex is added.
- The purpose of the consumer function. In general, side effect based functions make it hard to parse and understand, which makes it harder to maintain the code. If you can come up with a way to avoid that, it would be great. E.g you can make this function return all the vertices that were added to the graph, and then have the caller do what it needs to do by looping through those vertices. That is easier to parse than having to understand that this is a callback that is called on each vertex that was added to the graph.